packages/source-map/lib/sourceMap.ts

```diff
@@ -11,76 +11,151 @@ export interface Mapping<Data = unknown> {
+			const startMappingIndexes = startMatchesMap.get(endMapping[0])
+			if (startMappingIndexes && startMappingIndexes.has(endMapping[1])) {
+				perfectMatches.push(endMapping)
+			} else if (perfectMatches.length == 0) {
+				endMatches.push(endMapping)
+			}
+		}
+
+		const toRange: CodeRangeKey = fromRange == 'sourceOffsets' ? 'generatedOffsets' : 'sourceOffsets'
+		if (perfectMatches.length > 0) {
+			// Prefer the shortest mapping ranges
+			perfectMatches.sort((a, b) => getSpanLength(a, fromRange) - getSpanLength(b, fromRange))
+
+			for (const match of perfectMatches) {
+				const fromStartOffset = getStartOffset(match, fromRange)
+				const toStartOffset = getStartOffset(match, toRange)
+				const toLength = getSpanLength(match, toRange)
+				const fromLength = getSpanLength(match, fromRange)
+				if (toLength == fromLength) {
+					// `from` and `to` span mappings have the same length - map range verbatim
+					const startOffset = toStartOffset + start - fromStartOffset
+					yield [startOffset, startOffset + length, match[0]]
+				} else if (fromStartOffset == start && fromLength == length) {
+					// The whole `from` span is selected - map to the whole `to` span
+					yield [toStartOffset, toStartOffset + toLength, match[0]]
+				} else {
+					// We would need to do some heuristics here to map the span, and it would make little
+					// sense in terms of mapping accuracy.
+					// Try the next match.
 				}
-				skip.add(mapping);
+			}
+		} else {
+			// Prefer the shortest `from` spans
+			startMatches.sort((a, b) =>
+				getSpanLength(a, fromRange) - getSpanLength(b, fromRange))
+			endMatches.sort((a, b) =>
+				getSpanLength(a, fromRange) - getSpanLength(b, fromRange))
+
+			for (const startMatch of startMatches) {
+				let mapping = startMatch[0];
 
-				const mapped = translateOffset(offset, mapping[fromRange], mapping[toRange], getLengths(mapping, fromRange), getLengths(mapping, toRange));
-				if (mapped !== undefined) {
-					yield [mapped, mapping as Mapping<Data>] as const;
+				for (const endMatch of endMatches) {
+					if (endMatch[0] != mapping) {
+						continue
+					}
+					const startOffset = mapOffset(startMatch, start, fromRange, toRange)
+					const endOffset = mapOffset(endMatch, end, fromRange, toRange)
+					yield [startOffset, endOffset, mapping]
 				}
 			}
 		}
 	}
 
-	private getMemoBasedOnRange(fromRange: CodeRangeKey) {
+	private getStorageBasedOnRange(fromRange: CodeRangeKey) {
 		return fromRange === 'sourceOffsets'
-			? this.sourceCodeOffsetsMemo ??= this.createMemo('sourceOffsets')
-			: this.generatedCodeOffsetsMemo ??= this.createMemo('generatedOffsets');
+			? this.sourceCodeOffsetsStorage ??= this.createStorage('sourceOffsets')
+			: this.generatedCodeOffsetsStorage ??= this.createStorage('generatedOffsets');
+	}
+
+	private createStorage(key: CodeRangeKey): MapStorage<Data> {
+		if (this.mappings.every(mapping => areRangesSortedAndNonOverlapping(mapping[key], getLengths(mapping, key)))) {
```

Comment:
Several thoughts here:
1. We could also just throw an error if mappings are too large or contain overlapping ranges and not use `IntervalTreeStorage` at all. 
2. I wonder if there are any benefits to use `IntervalTreeStorage` in terms of performance.
3. It seems that we small modification, your structure can be reused for overlapping ranges as well. Which structure would be more performant?
4. The thing is, that if you have many mappings, storing their offsets and lengths in arrays takes much less memory as you don't need to create an object for each mapping - see the modified Angular template test how I use mappings.